### PR TITLE
Fix bug when resolver returns None and schema is a string

### LINKED ISSUE #566: Marshmallow plugin bug with string schema ref and schema_name_resolver
`resolve_schema_dict` claims to support `string` as its schema parameter type. It can get there for example when you are parsing schema from docstring using the provided yaml util, or using FlaskPlugin to do it for you. 

But, when `schema_name_resolver` returns `None` for this case, the schema gets forwarded to `schema2jsonschema` which expects a `marshmallow.Schema` instance and crashes.

Test case is a slightly modified "example application": 
<details>
<summary>Show example test case</summary>

```python
import uuid

from apispec import APISpec
from apispec.ext.marshmallow import MarshmallowPlugin
from apispec_webframeworks.flask import FlaskPlugin
from flask import Flask, jsonify
from marshmallow import fields, Schema


def schema_name_resolver(schema):
    if schema == "PetSchema":
        return None


# Create an APISpec
spec = APISpec(
    title="Swagger Petstore",
    version="1.0.0",
    openapi_version="3.0.2",
    plugins=[
        FlaskPlugin(),
        MarshmallowPlugin(schema_name_resolver=schema_name_resolver),
    ],
)


# Optional marshmallow support
class CategorySchema(Schema):
    id = fields.Int()
    name = fields.Str(required=True)


class PetSchema(Schema):
    categories = fields.List(fields.Nested(CategorySchema))
    name = fields.Str()


# Optional Flask support
app = Flask(__name__)


@app.route("/random")
def random_pet():
    """A cute furry animal endpoint.
    ---
    get:
      description: Get a random pet
      responses:
        200:
          description: Return a pet
          content:
            application/json:
              schema: PetSchema
    """
    # Hardcoded example data
    pet_data = {
        "name": "sample_pet_" + str(uuid.uuid1()),
        "categories": [{"id": 1, "name": "sample_category"}],
    }
    return PetSchema().dump(pet_data)


# Register the path and the entities within it
with app.test_request_context():
    spec.path(view=random_pet)


@app.route("/spec.json")
def get_spec():
    return jsonify(spec.to_dict())
```
</details>

<details>
<summary>Show example stacktrace</summary>

```
Traceback (most recent call last):
  File "/home//PycharmProjects/test-apispec/main.py", line 62, in <module>
    spec.path(view=random_pet)
  File "/home//.local/share/virtualenvs/test-apispec-HubT4B3L/lib/python3.8/site-packages/apispec/core.py", line 280, in path
    plugin.operation_helper(path=path, operations=operations, **kwargs)
  File "/home//.local/share/virtualenvs/test-apispec-HubT4B3L/lib/python3.8/site-packages/apispec/ext/marshmallow/__init__.py", line 202, in operation_helper
    self.resolver.resolve_response(response)
  File "/home//.local/share/virtualenvs/test-apispec-HubT4B3L/lib/python3.8/site-packages/apispec/ext/marshmallow/schema_resolver.py", line 113, in resolve_response
    self.resolve_schema(response)
  File "/home//.local/share/virtualenvs/test-apispec-HubT4B3L/lib/python3.8/site-packages/apispec/ext/marshmallow/schema_resolver.py", line 161, in resolve_schema
    content["schema"] = self.resolve_schema_dict(content["schema"])
  File "/home//.local/share/virtualenvs/test-apispec-HubT4B3L/lib/python3.8/site-packages/apispec/ext/marshmallow/schema_resolver.py", line 218, in resolve_schema_dict
    return self.converter.resolve_nested_schema(schema)
  File "/home//.local/share/virtualenvs/test-apispec-HubT4B3L/lib/python3.8/site-packages/apispec/ext/marshmallow/openapi.py", line 96, in resolve_nested_schema
    json_schema = self.schema2jsonschema(schema)
  File "/home//.local/share/virtualenvs/test-apispec-HubT4B3L/lib/python3.8/site-packages/apispec/ext/marshmallow/openapi.py", line 261, in schema2jsonschema
    fields = get_fields(schema)
  File "/home//.local/share/virtualenvs/test-apispec-HubT4B3L/lib/python3.8/site-packages/apispec/ext/marshmallow/common.py", line 63, in get_fields
    raise ValueError(
ValueError: 'PetSchema' doesn't have either `fields` or `_declared_fields`.
```
</details>

I think it should be enough to just pass `schema_instance` instead of `schema` to `self.schema2jsonschema` in `resolve_nested_schema`

### PR DESCRIPTION
**Pull request by Bangertm**
Fixes #566

### CODE REVIEW

File: tests/test_ext_marshmallow.py [hunk 1]
```diff
@@ -515,48 +515,54 @@ def test_schema_uses_ref_if_available_v3(self, spec_fixture):
             "schema"
         ] == build_ref(spec_fixture.spec, "schema", "Pet")
 
-    def test_schema_uses_ref_if_available_name_resolver_returns_none_v2(self):
+    @pytest.mark.parametrize(
+        "pet_schema", (PetSchema, PetSchema(), "tests.schemas.PetSchema"),
+    )
+    def test_schema_name_resolver_returns_none_v2(self, pet_schema):
```
Comment (Bangertm) [inline tests/test_ext_marshmallow.py:564]: These tests were previously not testing a unique code path. Changed them to expose this bug. We could add back the old tests as well but I'm not sure they were adding any value.

**Review by lafrech — COMMENTED** [review]
Thanks.

I totally agree with the fix.

The tests you're removing are non-reg tests introduced in #359. Maybe we introduced other tests incidentally testing the same thing since then but I'm not sure about dropping them. I doubt we'd reverse the logic introduced in #359 and break the case, but the same could be said about many non-reg test, so unless you're sure about it, I'd keep them.

Apart from this, fine for me.

**Comment by lafrech** [comment]
Let's not block this because of the test.

I pushed a new commit to your branch in which I reintroduce the tests that were removed. We can always remove them later if we deem them useless.

I also edited your commit to fix the name of the new test (test_schema_name_resolver..., "name" was missing in v2 test).

**Comment by lafrech** [comment]
@Bangertm I hope you don't mind me force-pushing in your branch. My time is limited these days and I didn't want to hold this more than needed.

**Comment by Bangertm** [comment]
No problem at all. I forgot about this thanks for finishing it.

**Pull request merged**